java-tools/OsmAndServer/src/main/java/net/osmand/server/api/services/ShareGpxService.java

```diff
@@ -0,0 +1,317 @@
+
+		if (sharingType == FileSharedInfo.SharingType.GROUP_BY_LOGIN && groupActions != null) {
+			String[] actions = groupActions.split(SEPARATOR);
+			for (String action : actions) {
+				FileSharedInfo.GroupAction groupAction = FileSharedInfo.GroupAction.fromAction(action.trim());
+				if (groupAction != null) {
+					fileSharedInfo.getGroupActions().add(groupAction);
+				}
+			}
+		}
+
+		return gson.toJsonTree(fileSharedInfo).getAsJsonObject();
+	}
+
+	public FileSharedInfo getSharedInfo(Long id) {
+		JsonObject sharedInfoJson = filesRepository.findSharedInfoById(id);
+		if (sharedInfoJson != null) {
+			return gson.fromJson(sharedInfoJson, FileSharedInfo.class);
+		}
+		return null;
+	}
+
+	public FileSharedInfo getSharedInfo(String code) {
+		JsonObject sharedInfoJson = filesRepository.findSharedInfoBySharedCode(code);
+		if (sharedInfoJson != null) {
+			return gson.fromJson(sharedInfoJson, FileSharedInfo.class);
+		}
+		return null;
+	}
+
+	public boolean checkAccess(FileSharedInfo info) {
+		FileSharedInfo.SharingType sharingType = info.getSharingType();
+		if (sharingType == FileSharedInfo.SharingType.PUBLIC) {
+			return true;
+		}
+		PremiumUserDevicesRepository.PremiumUserDevice dev = userdataService.checkUser();
+		if (dev == null) {
+			return false;
+		}
+		if (sharingType == FileSharedInfo.SharingType.PUBLIC_BY_LOGIN) {
+			return true;
+		}
+		if (sharingType == FileSharedInfo.SharingType.GROUP_BY_LOGIN) {
+			return !info.getBlacklist().getUsers().contains(dev.userid);
+		}
+		return false;
+	}
+
+	public ShareFileController.FileDownloadResult downloadFile(PremiumUserFilesRepository.UserFile userFile) throws IOException {
+		if (userFile == null) {
+			return null;
+		}
+		try (InputStream bin = userdataService.getInputStream(userFile)) {
+			if (bin != null) {
+				InputStream inputStream = new GZIPInputStream(bin);
+				String fileName = URLEncoder.encode(userdataService.sanitizeEncode(userFile.name), StandardCharsets.UTF_8);
+				return new ShareFileController.FileDownloadResult(inputStream, fileName, APPLICATION_OCTET_STREAM_VALUE);
+			}
+		}
+		return null;
+	}
+
+	public PremiumUserFilesRepository.UserFile getUserFileBySharedUrl(String token) {
+		return filesRepository.findUserFileBySharedCode(token);
+	}
+
+	public GpxFile getFile(PremiumUserFilesRepository.UserFile file) throws IOException {
+		if (file == null || file.data == null) {
+			return null;
+		}
+		try (InputStream inputStream = new GZIPInputStream(new ByteArrayInputStream(file.data));
+		     Source source = new Buffer().readFrom(inputStream)) {
+			GpxFile gpxFile = GpxUtilities.INSTANCE.loadGpxFile(source);
+			if (gpxFile.getError() == null) {
+				return gpxFile;
+			}
+			return null;
+		}
+	}
+
+	@Transactional
+	public void saveAccessedUser(PremiumUserDevicesRepository.PremiumUserDevice dev, PremiumUserFilesRepository.UserFile userFile, FileSharedInfo info) {
```

Comment:
Please avoid awkward naming. `accessed` is not adjective. Accessed user, last accessed time = incorrect usage. Consider logFileAccess instead of saveAccessedUser and generally change `accessedUsers` to `accessLog`.